cli/src/cmd/forge/script/cmd.rs

```diff
@@ -0,0 +1,167 @@
+use crate::cmd::{unwrap_contracts, Cmd, ScriptSequence};
+
+use ethers::{
+    abi::Abi,
+    prelude::{artifacts::CompactContractBytecode, ArtifactId},
+    types::{transaction::eip2718::TypedTransaction, U256},
+};
+use forge::executor::opts::EvmOpts;
+
+use foundry_config::{figment::Figment, Config};
+use foundry_utils::RuntimeOrHandle;
+
+use super::*;
+
+impl Cmd for ScriptArgs {
+    type Output = ();
+    fn run(self) -> eyre::Result<Self::Output> {
+        RuntimeOrHandle::new().block_on(self.run_script())
+    }
+}
+
+impl ScriptArgs {
+    pub async fn run_script(self) -> eyre::Result<()> {
+        let figment: Figment = From::from(&self);
+        let evm_opts = figment.extract::<EvmOpts>()?;
+        let mut script_config = ScriptConfig {
+            // dapptools compatibility
+            sender_nonce: U256::one(),
+            config: Config::from_provider(figment).sanitized(),
+            evm_opts,
+            called_function: None,
+        };
+
+        if let (Some(deployer), Some(fork_url)) =
+            (self.deployer, script_config.evm_opts.fork_url.as_ref())
+        {
+            script_config.sender_nonce = foundry_utils::next_nonce(deployer, fork_url, None)?
+        }
+
+        let BuildOutput {
+            project,
+            target,
+            contract,
+            highlevel_known_contracts,
+            predeploy_libraries,
+            known_contracts: default_known_contracts,
+            sources,
+        } = self.build(&script_config, self.deployer)?;
+
+        if self.force_resume || self.resume {
+            let mut deployment_sequence =
+                ScriptSequence::load(&self.sig, &target, &script_config.config.out)?;
+            self.send_transactions(&mut deployment_sequence).await?;
+        } else {
+            let mut known_contracts = unwrap_contracts(&highlevel_known_contracts);
+
+            // execute once with default sender
+            let mut result = self
+                .execute(&mut script_config, contract, self.deployer, &predeploy_libraries)
+                .await?;
+
+            let mut decoder = self.decode_traces(&script_config, &mut result, &known_contracts)?;
+
+            if self.debug {
+                self.run_debugger(&decoder, sources, result, project, highlevel_known_contracts)?;
+            } else {
+                if let Some(new_sender) = self.maybe_new_sender(
+                    &script_config.evm_opts,
+                    result.transactions.as_ref(),
+                    &predeploy_libraries,
```

Comment:
i've asked a couple of times but can @brockelmore or @joshieDo remind me why we need a new sender here

Reply:
when the user **does not define** a `--deployer` on the command line,  any predeployed libraries are first deployed locally using a random address. That happens because we can only know the signer/deployer once we execute the script and pass through a `vm.broadcast(address)`.

Here we just go through every collected transaction (if it exists and if there are predeploy libraries), and find the real sender.

Once that happens, we need to relink and execute everything again, because of `nonces`.

Reply:
Got it - could we document that part perhaps? So that it's clear that the motivation is "There may be multiple candidate addresses for deployment so [...]. In case there's 1 [..]".